bundles/org.openhab.core.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttBrokerConnection.java

```diff
@@ -604,30 +591,33 @@ public boolean hasSubscribers() {
      */
     @SuppressWarnings({ "null", "unused" })
     public CompletableFuture<Boolean> unsubscribe(String topic, MqttMessageSubscriber subscriber) {
+
+        final boolean needsUnsubscribe;
+
         synchronized (subscribers) {
-            final @Nullable List<MqttMessageSubscriber> list = subscribers.get(topic);
-            if (list == null) {
+            final @Nullable Subscription subscription = subscribers.get(topic);
+            if (subscription == null) {
                 logger.trace("Tried to unsubscribe {} from topic {}, but subscriber list is empty", subscriber, topic);
                 return CompletableFuture.completedFuture(true);
             }
-            list.remove(subscriber);
-            if (!list.isEmpty()) {
-                logger.trace("Removed {} from topic subscribers for topic {}, but other subscribers present",
-                        subscriber, topic);
-                return CompletableFuture.completedFuture(true);
+            subscription.remove(subscriber);
+
+            if (subscription.isEmpty()) {
+                needsUnsubscribe = true;
+                subscribers.remove(topic);
+            } else {
+                needsUnsubscribe = false;
             }
-            // Remove from subscriber list
-            subscribers.remove(topic);
-            // No more subscribers to this topic. Unsubscribe topic on the broker
+        }
+        if (needsUnsubscribe) {
```

Comment:
```suggestion
        if (subscribers.compute(topic, (t,v) -> v.isEmpty() ? null : v ) == null) {
```

and remove everything above?

Reply:
it is more complicated than that....
```
       if (subscribers.compute(topic, (t, v) -> {
            if (v != null) {
                v.remove(subscriber);
                if (v.isEmpty()) {
                    return null; // returning null here not allowed: required '@NonNull Subscription'
                }
            }
            return v;
        }) == null) {
```
And then it could be, that we try to unsubscribe, if we were never subscribed....

The question is actually: which is better to understand in a year....

Reply:
the remove should be before, I was only thinking of the removal from the map.

Reply:
The manipulation of the subscribers map needs to be atomic. Therefor put it intoa sync-block.
But the actucal call to the `subscribeRaw` I wanted outside the sync-block to keep the block as short as posssible.